File: CHAIN/source.cpp
/* Autor: Bartłomiej Romanek                  */
/* Tytuł: Wzorzec "Łańcuch odpowiedzialności" */
/* Licencja: Licencja MIT                     */
/* Data modyfikacji: 21.02.2017               */

#include <iostream>
#include <fstream>
#include <string>

using namespace std;

class Ekran {
	public:
		int Wyswietl(string & napis) {
			cout << napis << endl;
			return 0;
		}
};

class Plik {
	public:
		int Zapisz(string & napis) {
			ofstream plik("output.txt", ifstream::out);
			if (!plik.good()) {
				return 1;
			}
			plik << napis;
			plik.close();
			return 0;
		}
};

class Lancuch {
	public:
		void Zachowaj(string napis) {
        		preferencja = 0;
        		if (preferencja == 0) {
            			Plik file1;
            			preferencja = file1.Zapisz(napis);
             		}
             		if (preferencja == 1) {
    				Ekran display1;
    				preferencja = display1.Wyswietl(napis);
    			}	
		}
		
	private:
		int preferencja;
};

int main(){
	Lancuch strategy1;
	strategy1.Zachowaj("Bon na 10zl do Tesco: AH4S-C8H5-09NN");
	system("pause");
}

// Opis kodu: Klasa "Łancuch" posiada metodę "Zachowaj", która umożliwia zapis napisu do pliku. Jeżeli to się nie uda, 
// następuje próba wykonania operacji zastępczej: wypisanie napisu na ekran.

File: STRATEGY/source.php
/* Autor: Bartłomiej Romanek      */
/* Tytuł: Wzorzec "Strategia"     */
/* Licencja: Licencja MIT         */
/* Data modyfikacji: 28.02.2017   */

<?php

	class Ekran {
		public function Wyswietl($napis) {
			echo $napis . "\n";
		}
	};

	class Plik {
		public function Zapisz($napis) {
			$plik = fopen('output.txt', 'w');
			fwrite($plik, $napis);
			fclose($plik);
		}
	};

	class Strategia {
		public function Zachowaj($napis) {
			if ($this->preferencja == "ekran") {
				$display1 = new Ekran();
				$display1->Wyswietl($napis);
			}
			else if ($this->preferencja == "plik") {
				$file1 = new Plik();
				$file1->Zapisz($napis);
			}
			else {
				echo "Nie ustawiono strategii! \n";
			}	
		}
		public function Ustaw($opcja) {
			$this->preferencja = $opcja;
		}
		private $preferencja;
	};

	$strategy1 = new Strategia();
	$strategy1->Zachowaj("Bon na 10zl do Tesco: AH4S-C8H5-09NN");
	$strategy1->Ustaw("ekran");
	$strategy1->Zachowaj("Bon na 20zl do Tesco: AH4S-C8H5-09NN");
	$strategy1->Ustaw("plik");
	$strategy1->Zachowaj("Bon na 50zl do Tesco: AH4S-C8H5-09NN");

?>

// Opis kodu: Klasa "Strategia" posiada metodę "Zachowaj", która umożliwia przechowywanie napisu. W zależności od wybranej strategii 
// przez użytkownika (którą wybiera za pomocą metody "Ustaw"), zapis napisu wykonywany jest albo na ekran, albo do pliku.
